xapian-bindings/java/java.i

```diff
@@ -174,8 +174,11 @@ namespace Xapian {
     bool hasNext() const { return Xapian::iterator_valid(*self); }
 }
 
+%ignore MSetIterator::next();
+%javamethodmodifiers MSetIterator::next(int) "private"
+
 %extend MSetIterator {
-    Xapian::docid next () {
+    Xapian::docid next (int pseudoArg) {
```

Comment:
The whole work around starts here. We want to make MSetIterator itself be an iterator. So we make MSetIterator implement the `Iterator<MSetIterator.MatchDescriptor>` interface. To implement this interface, we need to implement two methods, hasNext() and next(). `public boolean hasNext()` is already implemented so no problem there. We also need to implement `public MatchDescriptor next()`, but here we face the problem since we already have a next() function. The only way to overcome this is to overload the next functions, where the main next function will be the one that is required by the interface. We overload by adding an argument to the function signature of `long next()`. So I added a pseudo argument which doesn't serve any purpose, it just enables us to overload.
`long next(int pseudoArg)` will then be a confusing method to deal with and it should be hidden outside of the class, so I made it a private method.

I feel this is the best way to make MSetIterator implement Iterator. The only issue here would be that `long next()` can't be used anymore. We could keep the method as public and explain in the documentation to the user that the argument should be kept as 0. If we keep it as private the only issue would be that next is being used differently in different languages.

@jaylett @ojwb What do you guys think? Is this a good way to do it or should I just wrap the MSetIterator into a custom implementation of Iterator interfaces?